pageserver/src/tenant/delete.rs

```diff
@@ -395,180 +241,191 @@ impl DeleteTimelineFlow {
+                {
                     error!("Error: {err:#}");
-                    timeline.set_broken(format!("{err:#}"))
+                    tenant.set_broken(err.to_string()).await;
                 };
                 Ok(())
             }
             .instrument({
-                let span =
-                    tracing::info_span!(parent: None, "delete_timeline", tenant_id=%tenant_id, timeline_id=%timeline_id);
+                let span = tracing::info_span!(parent: None, "delete_tenant", tenant_id=%tenant_id);
                 span.follows_from(Span::current());
                 span
             }),
         );
     }
 
     async fn background(
-        mut guard: DeletionGuard,
+        mut guard: OwnedMutexGuard<Self>,
         conf: &PageServerConf,
-        tenant: &Tenant,
-        timeline: &Timeline,
-    ) -> Result<(), DeleteTimelineError> {
-        delete_local_layer_files(conf, tenant.tenant_id, timeline).await?;
+        remote_storage: Option<GenericRemoteStorage>,
+        tenants: &'static tokio::sync::RwLock<TenantsMap>,
+        tenant: &Arc<Tenant>,
+    ) -> Result<(), DeleteTenantError> {
+        // Tree sort timelines, schedule delete for them. Mention retries from the console side.
+        // Note that if deletion fails we dont mark timelines as broken,
+        // the whole tenant will become broken as by `Self::schedule_background` logic
+        let already_running_timeline_deletions = schedule_ordered_timeline_deletions(tenant)
+            .await
+            .context("run_ordered_timeline_deletions")?;
+
+        // Wait for deletions that were already running at the moment when tenant deletion was requested.
+        // When we can lock deletion guard it means that corresponding timeline deletion finished.
+        for (guard, timeline_id) in already_running_timeline_deletions {
+            let flow = guard.lock().await;
+            if !flow.is_finished() {
+                return Err(DeleteTenantError::Other(anyhow::anyhow!(
+                    "already running timeline deletion failed: {timeline_id}"
+                )));
+            }
+        }
 
-        delete_remote_layers_and_index(timeline).await?;
+        assert_timeline_dir_empty(conf, tenant).await?;
 
-        pausable_failpoint!("in_progress_delete");
+        remove_tenant_remote_delete_mark(conf, remote_storage.as_ref(), &tenant.tenant_id).await?;
 
-        cleanup_remaining_timeline_fs_traces(conf, tenant.tenant_id, timeline.timeline_id).await?;
+        cleanup_remaining_fs_traces(conf, &tenant.tenant_id)
+            .await
+            .context("cleanup_remaining_fs_traces")?;
 
-        remove_timeline_from_tenant(tenant, timeline.timeline_id, &guard).await?;
+        let mut locked = tenants.write().await;
+        // NOTE:
+        // race with shutdown? -- not a problem because of completion::Barrier in TenantState::Stopping
+        // race with detach?
+        // race with ignore?
```

Comment:
Ok, thank you! Still using protection inside shutdown simultaneously with deletion guard may look a bit confusing. See also https://github.com/neondatabase/neon/pull/4855/files#diff-100b728a9529fbc86c8c054a1d7d3d0b5bc129a7ce27d28d833f60b7817169f7R338

Reply:
Do you have any recollection of where the see also could had pointed (the link is probably outdated).

Reply:
Hmm, yeah, the link is broken, points to nowhere. I think it was pointing to `Tenant::shutdown` call in `DeleteTenantFlow` here: https://github.com/zenithdb/zenith/blob/af2e26f0f4f1d630242e00e680934bf0ebf60ffc/pageserver/src/tenant/delete.rs#L437

Reply:
Reproduced:

```rust
// It would be good to only set stopping here and continue shutdown in the background, but shutdown is not idempotent.
// i e it is an error to do:
// tenant.set_stopping
// tenant.shutdown
// Its also bad that we're holding tenants.read here.
// TODO relax set_stopping to be idempotent?
if tenant.shutdown(progress, false).await.is_err() {
    return Err(DeleteTenantError::Other(anyhow::anyhow!(
        "tenant shutdown is already in progress"
    )));
}
```

> TODO: set_stopping to be idempotent

This sounds like a sensible idea to me, and then shutdown would be idempotent as there is no task to await for.

Unless in deletion case there might be the tasks to wait for, wouldn't this change that?

Reply:
Waiting for tasks should be also idempotent. If there are any -> wait for them if not -> continue. 

> Unless in deletion case there might be the tasks to wait for, wouldn't this change that?

If deletion fails I'd expect that it wont leave dangling tasks behind. Though this may not always hold. Since every upload queue op ends up in separate task they can outlive main deletion task. If we look at that from shutdown idempotency standpoint it should be fine to wait for such tasks to complete as well. 

I add this to follow up list. This is not critical at all, and currently nobody is calling delete from CP side

Reply:
Opened separate issue to track this https://github.com/neondatabase/neon/issues/4951